File: sorting/golang/quick.go
package main

import (
	"fmt"
)

func quick_sort(s []int) []int {
	return quick(s, 0, len(s)-1)
}

func quick(s []int, left int, right int) []int {
	if left >= right {
		return s
	}
	key := s[left]
	l, r := left, right
	for l < r {
		for s[r] > key && l < r {
			r -= 1
		}
		if l < r {
			s[l] = s[r]
			l += 1
		}
		for s[l] < key && l < r {
			l += 1
		}
		if l < r {
			s[r] = s[l]
			r -= 1
		}
	}
	s[l] = key
	quick(s, left, l-1)
	quick(s, r+1, right)
	return s
}

func main() {
	s := []int{1, 3, 9, 0, 2, 4, 7}
	fmt.Println(s, "---default")
	fmt.Println(quick_sort(s), "---quick")
}

File: sorting/python/quick.py
#!/usr/local/bin/python2
# -*- coding: UTF-8 -*-

def quick_sort(arr):
    return quick(arr, 0, len(arr) - 1)

def quick(arr, left, right):
    if left >= right:
        return arr
    l, r = left, right
    key = arr[left]
    while l < r:
        while l < r and arr[r] > key:
            r -= 1
        if l < r:
            arr[l] = arr[r]
            l += 1
        while l < r and arr[l] <= key:
            l += 1
        if l < r:
            arr[r] = arr[l]
            r -= 1
    arr[l] = key
    quick(arr, left, l - 1)
    quick(arr, r + 1, right)
    return arr

if __name__ == "__main__":
    arr = [1, 3, 9, 0, 2, 4, 7, 8, 6, 5]

    print arr, "---default"
    print quick_sort(arr), "---quick"
